src/psyclone/psyir/tools/definition_use_chains.py

```diff
@@ -621,3 +645,385 @@ def _find_basic_blocks(self, nodelist):
+                    ):
+                        stop_position = min(
+                            reference.abs_position, stop_position
+                        )
+        for region in basic_block_list:
+            region_list = region.walk((Reference, Call, CodeBlock, Return))
+            # Reverse the list
+            region_list.reverse()
+            for reference in region_list:
+                # Store the position instead of computing it twice.
+                abs_pos = reference.abs_position
+                if abs_pos < self._start_point or abs_pos >= stop_position:
+                    continue
+                if isinstance(reference, CodeBlock):
+                    # CodeBlocks only find symbols, so we can only do as good
+                    # as checking the symbol - this means we can get false
+                    # positives for structure accesses inside CodeBlocks.
+                    if isinstance(reference._fp2_nodes[0], Goto_Stmt):
+                        raise NotImplementedError(
+                            "DefinitionUseChains can't "
+                            "handle code containing GOTO"
+                            " statements."
+                        )
+                    if (
+                        self._reference.symbol.name
+                        in reference.get_symbol_names()
+                    ):
+                        # Assume the worst for a CodeBlock and we count them
+                        # as killed and defsout and uses.
+                        if defs_out is not None:
+                            self._killed.append(defs_out)
+                        defs_out = reference
+                        continue
+                elif isinstance(reference, Call):
+                    # If its a local variable we can ignore it as we'll catch
+                    # the Reference later if its passed into the Call.
+                    if self._reference.symbol.is_automatic:
+                        continue
+                    if isinstance(reference, IntrinsicCall):
+                        # IntrinsicCall can only do stuff to arguments, these
+                        # will be caught by Reference walk already.
+                        # Note that this assumes two symbols are not
+                        # aliases of each other.
+                        continue
+                    # For now just assume calls are bad if we have a non-local
+                    # variable and we treat them as though they were a write.
+                    if defs_out is not None:
+                        self._killed.append(defs_out)
+                    defs_out = reference
+                    continue
+                elif reference.get_signature_and_indices()[0] == sig:
+                    # Work out if its read only or not.
+                    assign = reference.ancestor(Assignment)
+                    # TODO Need to invert how we think about assignments.
```

Comment:
Is this TODO still necessary?

Reply:
Nope, this works now - removed the comment.